### PR TITLE
Add pico_get_unique_board_id_string API

### PR DESCRIPTION
**Pull request by earlephilhower**
Add a new API to pico_unique which will turn the unique ID into a
canonical text string.

Use this API to update the USB device serial number in stdio_usb.

Supersedes #280

### CODE REVIEW

File: src/rp2_common/pico_unique_id/unique_id.c [hunk 1]
```diff
@@ -26,3 +26,16 @@ static void __attribute__((constructor)) _retrieve_unique_id_on_boot(void) {
...
 }
+
+void pico_get_unique_board_id_string(char *id_out, size_t len) {
+    pico_unique_board_id_t temp = retrieved_id;
+    size_t i;
+    // Generate hex one nibble at a time
+    for (i = 0; (i < len - 1) && (i < PICO_UNIQUE_BOARD_ID_SIZE_BYTES * 2); i++) {
+        size_t bi = i / 2;
+        int nibble = (temp.id[bi] >> 4) & 0x0F;
+        temp.id[bi] = (uint8_t)(temp.id[bi] << 4);
```
Comment (lurch) [inline src/rp2_common/pico_unique_id/unique_id.c:37]: Apologies if this is a stupid question (I'm not a low-level C programmer), but is this modifying the bytes in the underlying `retrieved_id` ?

Comment (earlephilhower) [reply src/rp2_common/pico_unique_id/unique_id.c:37]: No, this works on the bytes in`temp` which are copied in line 31.

File: src/rp2_common/pico_stdio_usb/stdio_usb_descriptors.c [hunk 2]
```diff
@@ -121,6 +124,10 @@ const uint16_t *tud_descriptor_string_cb(uint8_t index, uint16_t langid) {
     #define DESC_STR_MAX (20)
     static uint16_t desc_str[DESC_STR_MAX];
 
+    // Assign the SN using the unique flash id
+    if (!usbd_serial_str[0])
```
Comment (kilograham) [inline src/rp2_common/pico_stdio_usb/stdio_usb_descriptors.c:128]: can you please use braces here. I guess we should check in a style guide

File: src/rp2_common/pico_unique_id/unique_id.c (hunk 1 above)
Comment (kilograham) [reply src/rp2_common/pico_unique_id/unique_id.c:37]: this does feel a bit weird though, i'd be tempted to just do
```
nibble = (temp.id[i/2] >> (4 * (i&1))) & 0xf;
```

File: src/rp2_common/pico_unique_id/unique_id.c [hunk 3]
```diff
@@ -26,3 +26,16 @@ static void __attribute__((constructor)) _retrieve_unique_id_on_boot(void) {
 void pico_get_unique_board_id(pico_unique_board_id_t *id_out) {
     *id_out = retrieved_id;
 }
+
+void pico_get_unique_board_id_string(char *id_out, size_t len) {
```
Comment (kilograham) [inline src/rp2_common/pico_unique_id/unique_id.c:30]: also for better or worse we've tended to use `uint` rather than `size_t` for sizes in APIs

**Review by kilograham — CHANGES_REQUESTED** [review]

File: src/rp2_common/pico_unique_id/unique_id.c (hunk 1 above)
Comment (lurch) [reply src/rp2_common/pico_unique_id/unique_id.c:37]: (I guess @kilograham 's suggestion doesn't modify `temp`, in which case perhaps there's no need for the copy and you can just access `retrieved_id`?)

Comment (earlephilhower) [reply src/rp2_common/pico_unique_id/unique_id.c:37]: That's a much better way and it doesn't need the copy, thanks.

File: src/rp2_common/pico_unique_id/unique_id.c (hunk 3 above)
Comment (earlephilhower) [reply src/rp2_common/pico_unique_id/unique_id.c]: All done.  To preserve hex ordering, we actually need to shift `(4 - 4 * (i&1))` or the hex nibbles end up backwards.

File: src/rp2_common/pico_unique_id/unique_id.c [hunk 4]
```diff
@@ -26,3 +26,14 @@ static void __attribute__((constructor)) _retrieve_unique_id_on_boot(void) {
 void pico_get_unique_board_id(pico_unique_board_id_t *id_out) {
     *id_out = retrieved_id;
 }
+
+void pico_get_unique_board_id_string(char *id_out, uint len) {
+    assert(len > 0);
```
Comment (lurch) [inline src/rp2_common/pico_unique_id/unique_id.c:31]: Isn't a uint _always_ > 0 ? :wink:

**Review by kilograham — APPROVED** [review]

File: src/rp2_common/pico_unique_id/unique_id.c (hunk 3 above)
Comment (lurch) [reply src/rp2_common/pico_unique_id/unique_id.c]: It's a good thing that `*` has a higher precedence than `-` :wink:

**Pull request merged**